pkg/replication/deployment_replication.go

```diff
@@ -0,0 +1,378 @@
+type deploymentReplicationEventType string
+
+const (
+	eventArangoDeploymentReplicationUpdated deploymentReplicationEventType = "DeploymentReplicationUpdated"
+)
+
+// seploymentReplicationEvent holds an event passed from the controller to the deployment replication.
+type deploymentReplicationEvent struct {
+	Type                  deploymentReplicationEventType
+	DeploymentReplication *api.ArangoDeploymentReplication
+}
+
+const (
+	deploymentReplicationEventQueueSize = 100
+	minInspectionInterval               = time.Second // Ensure we inspect the generated resources no less than with this interval
+	maxInspectionInterval               = time.Minute // Ensure we inspect the generated resources no less than with this interval
+)
+
+// DeploymentReplication is the in process state of an ArangoDeploymentReplication.
+type DeploymentReplication struct {
+	apiObject *api.ArangoDeploymentReplication // API object
+	status    api.DeploymentReplicationStatus  // Internal status of the CR
+	config    Config
+	deps      Dependencies
+
+	eventCh chan *deploymentReplicationEvent
+	stopCh  chan struct{}
+	stopped int32
+
+	eventsCli corev1.EventInterface
+
+	inspectTrigger         trigger.Trigger
+	recentInspectionErrors int
+	clientCache            client.ClientCache
+}
+
+// New creates a new DeploymentReplication from the given API object.
+func New(config Config, deps Dependencies, apiObject *api.ArangoDeploymentReplication) (*DeploymentReplication, error) {
+	if err := apiObject.Spec.Validate(); err != nil {
+		return nil, maskAny(err)
+	}
+	dr := &DeploymentReplication{
+		apiObject: apiObject,
+		status:    *(apiObject.Status.DeepCopy()),
+		config:    config,
+		deps:      deps,
+		eventCh:   make(chan *deploymentReplicationEvent, deploymentReplicationEventQueueSize),
+		stopCh:    make(chan struct{}),
+		eventsCli: deps.KubeCli.Core().Events(apiObject.GetNamespace()),
+	}
+
+	go dr.run()
+
+	return dr, nil
+}
+
+// Update the deployment replication.
+// This sends an update event in the event queue.
+func (dr *DeploymentReplication) Update(apiObject *api.ArangoDeploymentReplication) {
+	dr.send(&deploymentReplicationEvent{
+		Type: eventArangoDeploymentReplicationUpdated,
+		DeploymentReplication: apiObject,
+	})
+}
+
+// Delete the deployment replication.
+// Called when the local storage was deleted by the user.
+func (dr *DeploymentReplication) Delete() {
+	dr.deps.Log.Info().Msg("deployment replication is deleted by user")
+	if atomic.CompareAndSwapInt32(&dr.stopped, 0, 1) {
+		close(dr.stopCh)
+	}
+}
+
+// send given event into the deployment replication event queue.
+func (dr *DeploymentReplication) send(ev *deploymentReplicationEvent) {
+	select {
+	case dr.eventCh <- ev:
+		l, ecap := len(dr.eventCh), cap(dr.eventCh)
+		if l > int(float64(ecap)*0.8) {
+			dr.deps.Log.Warn().
+				Int("used", l).
+				Int("capacity", ecap).
+				Msg("event queue buffer is almost full")
+		}
+	case <-dr.stopCh:
```

Comment:
go ignorance here, if dr.stopCh has closed will ev still go to queue and all is ok?

Reply:
Yes. The cases of a select statement are tried in order.

Reply:
Of course it can happen that ev cannot be put into the queue (it is full) and stopCh is closed. In that case the stopCh case will win (as intended)